File: mobile/lib/data/community/group_member_tour.dart
import '../../domain/saved_route.dart';
import '../../domain/saved_route_note.dart';
import '../../domain/tours/tour_akte.dart';
import 'ride_group_invite.dart';
import 'tour_share_codec.dart';

/// Invite-Token-Limit inkl. eingebetteter Spur (WhatsApp/SMS).
const maxGroupInviteTokenChars = 2400;

/// Private Host-GPX: Spur ins Invite, nicht auf Explore.
Map<String, dynamic>? tourShareForGroupInvite({
  required String savedRouteId,
  String? catalogTourId,
  SavedRouteEntry? route,
  SavedRouteMeta? meta,
}) {
  if (route == null) return null;
  if (!needsMemberTrack(
    savedRouteId: savedRouteId,
    catalogTourId: catalogTourId ?? meta?.catalogTourId,
  )) {
    return null;
  }
  return buildTourSharePayload(route, meta: meta ?? SavedRouteMeta.empty);
}

/// Gast-Kopie mit Host-Id, damit Losfahren matcht. Ohne Spur: null.
SavedRouteEntry? importMemberTourFromInvite({
  required RideGroupInvitePayload? payload,
  required List<SavedRouteEntry> existing,
}) {
  if (payload == null) return null;
  if (!needsMemberTrack(
    savedRouteId: payload.savedRouteId,
    catalogTourId: payload.catalogTourId,
  )) {
    return null;
  }
  if (existing.any((r) => r.id == payload.savedRouteId)) return null;
  final tour = payload.tour;
  if (tour == null) return null;
  return savedRouteFromTourShare(
    tour: tour,
    keepId: payload.savedRouteId,
  );
}

File: src/lib/community/groupMemberTour.ts
import { needsMemberTrack } from "@/lib/community/rideGroup";
import type { RideGroup } from "@/lib/community/types";
import type { RideGroupInvitePayload } from "@/lib/community/rideGroupInvite";
import {
  buildTourSharePayload,
  parseTourShareMap,
  savedRouteFromTourShare,
} from "@/lib/community/shareCodec";
import type { SharedTourPayload } from "@/lib/community/shareTypes";
import type { SavedRoute } from "@/types/route";

/** Invite-Token-Limit inkl. eingebetteter Spur (WhatsApp/SMS). */
export const MAX_GROUP_INVITE_TOKEN_CHARS = 2400;

/** Private Host-GPX: Spur ins Invite, nicht auf Explore. */
export function tourShareForGroupInvite(input: {
  savedRouteId: string;
  catalogTourId?: string | null;
  route?: SavedRoute | null;
}): SharedTourPayload | undefined {
  const route = input.route;
  if (!route) return undefined;
  if (
    !needsMemberTrack({
      savedRouteId: input.savedRouteId,
      catalogTourId: input.catalogTourId ?? route.catalogTourId,
    })
  ) {
    return undefined;
  }
  return buildTourSharePayload(route);
}

/** Gast-Kopie mit Host-Id, damit Losfahren matcht. Ohne Spur: null. */
export function importMemberTourFromInvite(input: {
  payload?: RideGroupInvitePayload | null;
  existing: SavedRoute[];
}): SavedRoute | null {
  const payload = input.payload;
  if (!payload) return null;
  if (
    !needsMemberTrack({
      savedRouteId: payload.savedRouteId,
      catalogTourId: payload.catalogTourId,
    })
  ) {
    return null;
  }
  if (input.existing.some((r) => r.id === payload.savedRouteId)) return null;
  const tour = parseTourShareMap(payload.tour);
  if (!tour) return null;
  return savedRouteFromTourShare(tour, payload.savedRouteId);
}

export function tourForInviteGroup(
  group: Pick<RideGroup, "savedRouteId" | "catalogTourId">,
  route?: SavedRoute | null
): SharedTourPayload | undefined {
  return tourShareForGroupInvite({
    savedRouteId: group.savedRouteId,
    catalogTourId: group.catalogTourId,
    route,
  });
}
